File: problems/minimum_flips_in_binary_tree_to_get_result/minimum_flips_in_binary_tree_to_get_result_test.py
"""Test cases for the minimumFlips function."""

import sys
from pathlib import Path

import pytest

# Add the root directory to Python path
root_dir = Path(__file__).parent.parent.parent
sys.path.insert(0, str(root_dir))

from common import TreeNode  # noqa: E402
from minimum_flips_in_binary_tree_to_get_result import minimumFlips  # noqa: E402


@pytest.mark.parametrize(
    "root, result, expected",
    [
        (TreeNode(0), True, 1),  # Flip leaf node from 0 to 1
        (TreeNode(1), True, 0),  # Leaf node already 1
        (TreeNode(2, TreeNode(0), TreeNode(0)), True, 1),  # OR gate, flip one child
        (
            TreeNode(2, TreeNode(1), TreeNode(1)),
            False,
            2,
        ),  # OR gate, flip both children
        (TreeNode(3, TreeNode(1), TreeNode(1)), True, 0),  # AND gate, already true
        (TreeNode(3, TreeNode(1), TreeNode(0)), True, 1),  # AND gate, flip one child
        (TreeNode(4, TreeNode(1), TreeNode(0)), True, 0),  # XOR gate, already true
        (TreeNode(4, TreeNode(1), TreeNode(1)), False, 0),  # XOR gate, already false
        (TreeNode(5, TreeNode(0)), True, 0),  # NOT gate, already true
        (TreeNode(5, TreeNode(1)), False, 0),  # NOT gate, already false
        (
            TreeNode(
                2, TreeNode(3, TreeNode(1), TreeNode(0)), TreeNode(5, TreeNode(0))
            ),
            True,
            0,
        ),  # Complex tree: OR(AND(1,0), NOT(0)) = 1, already true
        (
            TreeNode(
                3, TreeNode(2, TreeNode(0), TreeNode(0)), TreeNode(5, TreeNode(1))
            ),
            False,
            0,
        ),  # Complex tree: AND(OR(0,0), NOT(1)) = 0, already false
        (
            TreeNode(
                4,
                TreeNode(2, TreeNode(1), TreeNode(0)),
                TreeNode(3, TreeNode(1), TreeNode(1)),
            ),
            True,
            1,
        ),  # Complex tree: XOR(OR(1,0), AND(1,1)) = 1, need 1 flip to make false
        (
            TreeNode(
                4,
                TreeNode(2, TreeNode(1), TreeNode(1)),
                TreeNode(3, TreeNode(1), TreeNode(0)),
            ),
            False,
            1,
        ),  # Complex tree: XOR(OR(1,1), AND(1,0)) = 1, flip one child to make it 0
        (
            TreeNode(
                4,
                TreeNode(0),
                TreeNode(
                    2, TreeNode(0), TreeNode(5, None, TreeNode(5, None, TreeNode(1)))
                ),
            ),
            True,
            0,
        ),  # Complex tree: XOR(0, OR(0, NOT(NOT(1)))) = 1, already true
    ],
    ids=[
        "leaf_0_to_true",
        "leaf_1_already_true",
        "or_gate_flip_child",
        "or_gate_to_false",
        "and_gate_already_true",
        "and_gate_flip_child",
        "xor_gate_already_true",
        "xor_gate_already_false",
        "not_gate_already_true",
        "not_gate_already_false",
        "complex_tree_or_already_true",
        "complex_tree_and_already_false",
        "complex_tree_xor_need_flip",
        "complex_tree_xor_to_false",
        "complex_xor_tree_with_not_right_child",
    ],
)
def test_minimum_flips(root, result, expected):
    assert minimumFlips(root, result) == expected

File: problems/minimum_flips_in_binary_tree_to_get_result/minimum_flips_in_binary_tree_to_get_result_test.cc
#include "minimum_flips_in_binary_tree_to_get_result.h"

#include <gtest/gtest.h>
#include <string>
#include "../../common/trees/treenode.h"

struct MinimumFlipsCase {
    std::string test_name;
    TreeNode *root;
    bool result;
    int expected;
};

using MinimumFlipsTest = ::testing::TestWithParam<MinimumFlipsCase>;

TEST_P(MinimumFlipsTest, TestCases) {
    const MinimumFlipsCase &testCase = GetParam();
    const int result = minimumFlips(testCase.root, testCase.result);
    EXPECT_EQ(result, testCase.expected);
}

INSTANTIATE_TEST_SUITE_P(
    MinimumFlipsTestCases, MinimumFlipsTest,
    ::testing::Values(
        MinimumFlipsCase{
            .test_name = "Leaf0ToTrue", .root = new TreeNode(0), .result = true, .expected = 1},
        MinimumFlipsCase{.test_name = "Leaf1AlreadyTrue",
                         .root = new TreeNode(1),
                         .result = true,
                         .expected = 0},
        MinimumFlipsCase{.test_name = "OrGateFlipChild",
                         .root = new TreeNode(2, new TreeNode(0), new TreeNode(0)),
                         .result = true,
                         .expected = 1},
        MinimumFlipsCase{.test_name = "OrGateToFalse",
                         .root = new TreeNode(2, new TreeNode(1), new TreeNode(1)),
                         .result = false,
                         .expected = 2},
        MinimumFlipsCase{.test_name = "AndGateAlreadyTrue",
                         .root = new TreeNode(3, new TreeNode(1), new TreeNode(1)),
                         .result = true,
                         .expected = 0},
        MinimumFlipsCase{.test_name = "AndGateFlipChild",
                         .root = new TreeNode(3, new TreeNode(1), new TreeNode(0)),
                         .result = true,
                         .expected = 1},
        MinimumFlipsCase{.test_name = "XorGateAlreadyTrue",
                         .root = new TreeNode(4, new TreeNode(1), new TreeNode(0)),
                         .result = true,
                         .expected = 0},
        MinimumFlipsCase{.test_name = "XorGateAlreadyFalse",
                         .root = new TreeNode(4, new TreeNode(1), new TreeNode(1)),
                         .result = false,
                         .expected = 0},
        MinimumFlipsCase{.test_name = "NotGateAlreadyTrue",
                         .root = new TreeNode(5, new TreeNode(0), nullptr),
                         .result = true,
                         .expected = 0},
        MinimumFlipsCase{.test_name = "NotGateAlreadyFalse",
                         .root = new TreeNode(5, new TreeNode(1), nullptr),
                         .result = false,
                         .expected = 0},
        MinimumFlipsCase{.test_name = "ComplexTreeOrAlreadyTrue",
                         .root = new TreeNode(2, new TreeNode(3, new TreeNode(1), new TreeNode(0)),
                                              new TreeNode(5, new TreeNode(0), nullptr)),
                         .result = true,
                         .expected = 0},
        MinimumFlipsCase{.test_name = "ComplexTreeAndAlreadyFalse",
                         .root = new TreeNode(3, new TreeNode(2, new TreeNode(0), new TreeNode(0)),
                                              new TreeNode(5, new TreeNode(1), nullptr)),
                         .result = false,
                         .expected = 0},
        MinimumFlipsCase{.test_name = "ComplexTreeXorNeedFlip",
                         .root = new TreeNode(4, new TreeNode(2, new TreeNode(1), new TreeNode(0)),
                                              new TreeNode(3, new TreeNode(1), new TreeNode(1))),
                         .result = true,
                         .expected = 1},
        MinimumFlipsCase{.test_name = "ComplexTreeXorToFalse",
                         .root = new TreeNode(4, new TreeNode(2, new TreeNode(1), new TreeNode(1)),
                                              new TreeNode(3, new TreeNode(1), new TreeNode(0))),
                         .result = false,
                         .expected = 1},
        MinimumFlipsCase{
            .test_name = "Leaf0ToFalse", .root = new TreeNode(0), .result = false, .expected = 0},
        MinimumFlipsCase{
            .test_name = "Leaf1ToFalse", .root = new TreeNode(1), .result = false, .expected = 1},
        MinimumFlipsCase{.test_name = "AndGateToFalse",
                         .root = new TreeNode(3, new TreeNode(0), new TreeNode(1)),
                         .result = false,
                         .expected = 0},
        MinimumFlipsCase{.test_name = "NotGateFlipToTrue",
                         .root = new TreeNode(5, new TreeNode(1), nullptr),
                         .result = true,
                         .expected = 1},
        MinimumFlipsCase{.test_name = "NotGateFlipToFalse",
                         .root = new TreeNode(5, new TreeNode(0), nullptr),
                         .result = false,
                         .expected = 1},
        MinimumFlipsCase{
            .test_name = "ComplexXorTreeWithNotRightChild",
            .root = new TreeNode(
                4, new TreeNode(0),
                new TreeNode(2, new TreeNode(0),
                             new TreeNode(5, nullptr, new TreeNode(5, nullptr, new TreeNode(1))))),
            .result = true,
            .expected = 0}),
    [](const ::testing::TestParamInfo<MinimumFlipsCase> &info) { return info.param.test_name; });